apsuite/commisslib/measure_bbb_data.py

```diff
@@ -523,3 +523,250 @@ def waterfall_plot(axis, xs, zs, data):
+        self.turn_on_pingers()
+        self.turn_on_pingers_pulse()
+
+    def inject_in_storage_ring(self, goal_curr):
+        """."""
+        self.devices['egun'].cmd_enable_trigger()
+        while not self._check_stored_current(goal_curr):
+            _time.sleep(TuneShiftParams.WAIT_INJ)
+        curr = self.devices['bbbh'].dcct.current
+        print(
+            f'Stored Current: {curr:.3f}/{goal_curr:.3f}mA.')
+        self.devices['egun'].cmd_disable_trigger()
+
+    def _check_stored_current(
+            self, goal_curr, tol=TuneShiftParams.DEF_TOL_CURRENT):
+        dcct_curr = self.devices['bbbh'].dcct.current
+        return dcct_curr > goal_curr or abs(dcct_curr - goal_curr) < tol
+
+    def _check_pingers_problem(self):
+        for ping in self.pingers:
+            if ping.voltage_mon < 0:
+                # reset pinger
+                ping.cmd_turn_off()
+                ping.cmd_turn_on()
+                return True
+        return False
+
+    def run_meas(self, save=True):
+        """."""
+        data = dict()
+        datah = list()
+        datav = list()
+        currs = list()
+        for curr in self.params.currents:
+            t0 = _time.time()
+            self.inject_in_storage_ring(curr)
+
+            trial = 0
+            while self._check_pingers_problem():
+                if trial > 2:
+                    print('3 unsucessful reset trials. Exiting...')
+                    break
+                print('Problem with pingers voltage. Resetting...')
+                trial += 1
+
+            _time.sleep(self.params.wait_bbb)
+            print('Acquiring data...')
+            currs.append(self.devices['bbbh'].dcct.current)
+            data['stored_current'] = currs
+            if 'H' in self.params.plane:
+                datah.append(self.get_data(plane='H'))
+                data['horizontal'] = datah
+            if 'V' in self.params.plane:
+                datav.append(self.get_data(plane='V'))
+                data['vertical'] = datav
+            self.data = data
+            if save:
+                self.save_data(fname=self.params.filename, overwrite=True)
+                print('Data saved!')
+            tf = _time.time()
+            print(f'Elapsed time: {tf-t0:.2f}s \n')
+
+    def plot_spectrum(
+            self, plane, freq_min=None, freq_max=None,
+            title=None, fname=None):
+        """plane: must be 'H' or 'V'."""
+        fig = _mplt.figure(figsize=(8, 6))
+        gs = _mgs.GridSpec(1, 1)
+        ax = fig.add_subplot(gs[0, 0])
+
+        curr = _np.array(self.data['stored_current'])
+        if plane.upper() == 'H':
+            data = self.data['horizontal']
+            freq_min = freq_min or 38
+            freq_max = freq_max or 52
+        if plane.upper() == 'V':
+            data = self.data['vertical']
+            freq_min = freq_min or 72
+            freq_max = freq_max or 84
+
+        mag = list()
```

Comment:
maybe: `mag = [dt['spec_mag'] for dt in data]` instead of the loop?

Reply:
yes, much better!